qcodes/dataset/data_set_cache.py

```diff
@@ -299,7 +311,10 @@ def _merge_data(existing_data: Mapping[str, np.ndarray],
                 ) -> Tuple[Dict[str, np.ndarray], Optional[int]]:
 
     subtree_merged_data = {}
-    subtree_parameters = set(existing_data.keys()) | set(new_data.keys())
+    subtree_parameters = existing_data.keys()
+
+    assert set(new_data.keys()).issubset(set(existing_data.keys()))
```

Comment:
Should probably just convert this to a proper error.

Reply:
To prevent this from ever happening we should probably require that prepare is always called on the dataset. (see the last tests that I just had to fix since they did not call prepare but the lower lying api)

Reply:
Rewrote this so cache is prepared at the same step that marks the dataset non pristine so we avoid additional logic